practice/tic-tac-toe/dict_board.py

```diff
@@ -0,0 +1,185 @@
+
+        (top-left is 0, 0), x is horizontal position, y is vertical position
+
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {(0, 0): 'X'}
+        >>> a.place_token(0, 1, 'O')
+        >>> b = a._coord_to_token
+        >>> sorted(b.items())
+        [((0, 0), 'X'), ((0, 1), 'O')]
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {(0, 0): 'X', (0,1): 'O'}
+        >>> a.place_token(0, 2, 'X')
+        >>> b = a._coord_to_token
+        >>> sorted(b.items())
+        [((0, 0), 'X'), ((0, 1), 'O'), ((0, 2), 'X')]
+        """
+        self._coord_to_token[(x, y)] = token
+
+    def calc_winner(self):
+        """Determines what token string has won or returns None if no one has
+
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {}
+        >>> a.calc_winner() == None
+        True
+
+
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {
+        ... (0, 0): 'X', (1, 0): 'X', (2, 0): '0',
+        ... (0, 1): 'O', (1, 1): 'O', (2, 1): 'X',
+        ... (0, 2): 'X', (1, 2): 'O', (2, 2): 'X'
+        ... }
+        >>> a.calc_winner() == None
+        True
+
+
+             >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {
+        ... (0, 2): 'X', (1, 2): 'X', (2, 2): 'X'
+        ... }
+        >>> a.calc_winner()
+        'X'
+
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {
+        ... (2, 0): 'O',
+        ... (2, 1): 'O',
+        ... (2, 2): 'O'
+        ... }
+        >>> a.calc_winner()
+        'O'
+
+        >>> a = DictTTTBoard()
+        >>> a._coord_to_token = {
+        ... (0, 0): 'X', (2, 0): 'O',
+        ... (1, 1): 'X',
+        ... (2, 2): 'X'
+        ... }
+        >>> a.calc_winner()
+        'X'
+        """
+        win_keys = [
+            [self._coord_to_token.get((1, 0), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((1, 2), ' ')],
+
+            [self._coord_to_token.get((0, 1), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((2, 1), ' ')],
+
+            [self._coord_to_token.get((0, 2), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((2, 0), ' ')],
+
+            [self._coord_to_token.get((0, 0), ' '), self._coord_to_token.get((1, 1), ' '),
+             self._coord_to_token.get((2, 2), ' ')],
+
+            [self._coord_to_token.get((0, 0), ' '), self._coord_to_token.get((1, 2), ' '),
+             self._coord_to_token.get((0, 2), ' ')],
+
+            [self._coord_to_token.get((0, 0), ' '), self._coord_to_token.get((0, 1), ' '),
+             self._coord_to_token.get((0, 2), ' ')],
+
+            [self._coord_to_token.get((2, 2), ' '), self._coord_to_token.get((2, 1), ' '),
+             self._coord_to_token.get((2, 0), ' ')],
+
+            [self._coord_to_token.get((2, 2), ' '), self._coord_to_token.get((1, 2), ' '),
+             self._coord_to_token.get((0, 2), ' ')]
+        ]
```

Comment:
There is also a significant textual pattern here. Have the computer generate the pattern.

```py
row_win_keys = ...
col_win_keys = ...
diag_win_keys = ...
all_win_keys = row_win_keys + col_win_keys + diag_win_keys
```